### PR TITLE
Routine to create dplms dictionary for Ge processing

### PR DESCRIPTION
**Pull request by valerioda**
Created new pargen rountine for DPLMS filter creation, this works by taking FFT and CAL files:

- noise matrix using selected baseline events
- signal matrix using selected FEP events
- filter synthesis by solving linear system
- small grid search by changing penalized coefficients

### CODE REVIEW

File: src/pygama/pargen/energy_optimisation.py [hunk 1]
```diff
@@ -1216,39 +1217,37 @@ def single_peak_fom(data, kwarg_dict):
     return out_dict
 
 
-def new_fom(data, kwarg_dict):
+def new_fom(data, kwarg_dict, alpha=None):
```
Comment (ggmarshall) [inline src/pygama/pargen/energy_optimisation.py:1220]: I think adding an alpha here would break the dsp_optimize routines, would need checking

File: src/pygama/pargen/energy_optimisation.py [hunk 2]
```diff
@@ -1084,19 +1084,19 @@ def event_selection(
     return final_data, idx_list
 
 
-def fwhm_slope(x, m0, m1, m2):
+def fwhm_slope(x, m0, m1):
```
Comment (ggmarshall) [inline src/pygama/pargen/energy_optimisation.py:1087]: maybe make m2 optional, keyword argument default None?

File: src/pygama/pargen/energy_optimisation.py [hunk 3]
```diff
@@ -954,16 +956,14 @@ def event_selection(
...
+    dEuc = 1 / guess_keV
     hist, bins, var = pgh.get_hist(rough_energy, range=(Euc_min, Euc_max), dx=dEuc)
     detected_peaks_locs, detected_peaks_keV, roughpars = pgc.hpge_find_E_peaks(
-        hist,
-        bins,
-        var,
-        np.array([238.632, 583.191, 727.330, 860.564, 1620.5, 2103.53, 2614.553]),
+        hist, bins, var, peaks_keV, n_sigma=3
     )
+    print(f"detected {detected_peaks_keV} keV peaks at {detected_peaks_locs}")
```
Comment (ggmarshall) [inline src/pygama/pargen/energy_optimisation.py:966]: remove print statement

File: src/pygama/pargen/energy_optimisation.py [hunk 4]
```diff
@@ -942,6 +943,7 @@ def event_selection(
                 else:
                     final_mask = final_mask | e_cut
             ids = final_mask
+            print(f"pulser found: {pulser_props}")
```
Comment (ggmarshall) [inline src/pygama/pargen/energy_optimisation.py:946]: remove

File: src/pygama/pargen/energy_optimisation.py [hunk 5]
```diff
@@ -954,16 +956,14 @@ def event_selection(
...
     Euc_max = 2620 / guess_keV * 1.1
-    dEuc = 1  # / guess_keV
+    dEuc = 1 / guess_keV
     hist, bins, var = pgh.get_hist(rough_energy, range=(Euc_min, Euc_max), dx=dEuc)
     detected_peaks_locs, detected_peaks_keV, roughpars = pgc.hpge_find_E_peaks(
-        hist,
-        bins,
-        var,
-        np.array([238.632, 583.191, 727.330, 860.564, 1620.5, 2103.53, 2614.553]),
+        hist, bins, var, peaks_keV, n_sigma=3
```
Comment (ggmarshall) [inline src/pygama/pargen/energy_optimisation.py:964]: Why are these changes needed? They will break other routines

File: src/pygama/pargen/energy_optimisation.py (hunk 1 above)
Comment (valerioda) [reply src/pygama/pargen/energy_optimisation.py:1220]: default is None, so should not impact other routines.
I added this to allow calling `new_fom` having already an alpha value

**Comment by gipert** [comment]
Valerio could you please rebase this PR on `main` and update the LH5 file writes/reads to match the new `LH5Store` syntax?

File: src/pygama/pargen/dplms_ge_dict.py [hunk 6]
```diff
@@ -0,0 +1,575 @@
...
+import json
+import logging
+import os
+import pathlib
+import pickle
+import time
+from collections import OrderedDict
+
+import lgdo
+import lgdo.lh5_store as lh5
```
Comment (gipert) [inline src/pygama/pargen/dplms_ge_dict.py:17]:
```suggestion
from lgdo import lh5
```

File: src/pygama/pargen/dplms_ge_dict.py [hunk 7]
```diff
@@ -0,0 +1,575 @@
...
+    decay_const: float = 0,
+    ene_par: str = "dplmsEmax",
+    display: int = 0,
+) -> dict:
+    """
+    This function calculates the dplms dictionary for HPGe detectors.
+
+    Parameters
+    ----------
+    lh5_path: str
```
Comment (gipert) [inline src/pygama/pargen/dplms_ge_dict.py:60]: If you use type annotations in the function signature, there is no need to duplicate them in the docstring:

```suggestion
    lh5_path,
```

File: src/pygama/pargen/dplms_ge_dict.py [hunk 8]
```diff
@@ -0,0 +1,575 @@
...
+        nm_coeff * nmat,
+        rmat,
+        za_coeff,
+        pmat,
+        ft_coeff * fmat,
+        dplms_dict["length"],
+        wsize,
+    )
+
+    sto.write_object(
```
Comment (gipert) [inline src/pygama/pargen/dplms_ge_dict.py:278]:
```suggestion
    sto.write(
```

**Review by gipert — CHANGES_REQUESTED** [review]

**Review by valerioda — COMMENTED** [review]
update the LH5 file writes/reads to match the new LH5Store syntax according suggestions

**Comment by gipert** [comment]
Should I merge then?

**Comment by valerioda** [comment]
> Should I merge then?

after doing some change, for me is ready to be merged

**Review by gipert — APPROVED** [review]
